File: collect.py
#!/usr/bin/env python3
"""
Cognito Forms access collector, browser-driven.

Opens Chromium against a saved profile directory so you sign in once and the
session sticks. Runs the same read-only collection the console script does,
writes audit-data.js next to report.html, then opens the report.

    python collect.py --org yourorgcode

Cognito resolves access as a global tier plus sparse overrides, so this keeps
every override row including denials. Everything it touches is a read.
"""

import argparse
import json
import pathlib
import sys
import webbrowser

try:
    from playwright.sync_api import sync_playwright
except ImportError:
    sys.exit("Playwright is not installed. Run:\n"
             "  pip install -r requirements.txt\n"
             "  playwright install chromium")

ROOT = pathlib.Path(__file__).resolve().parent
PROFILE_DIR = ROOT / ".browser-profile"
OUT_FILE = ROOT / "audit-data.js"
REPORT = ROOT / "report.html"

# Same three endpoints as collect.js, plus the archived-form resolution pass.
# That version saves through the browser; this one returns the payload to
# Python. If you change one, change both.
COLLECT = """
() => {
  const CONFIG = { orgId: %s };
  return (async () => {
    const orgCode = location.pathname.split('/')[1] || '';

    const metaHeaders = orgId => {
      const h = { 'content-type': 'application/json', 'x-requested-with': 'XMLHttpRequest' };
      if (orgId) h['x-organization'] = orgId;
      return h;
    };
    async function meta(orgId, formIds) {
      const r = await fetch('/svc/entity-meta/forms', {
        method: 'POST', headers: metaHeaders(orgId), credentials: 'include',
        body: JSON.stringify({ formIds: formIds || [] })
      });
      if (!r.ok) return null;
      try { return await r.json(); } catch (e) { return null; }
    }

    const roster = await (await fetch('/svc/organization/users', { credentials: 'include' })).json();
    if (!Array.isArray(roster.Users) || !roster.Users.length) throw new Error('roster empty');

    let m = await meta(null, []);
    let usedOrgId = null;
    if (!m || !(m.Forms || []).length) { m = await meta(CONFIG.orgId, []); usedOrgId = CONFIG.orgId; }
    if (!m || !(m.Forms || []).length) throw new Error('entity-meta returned no forms');

    // Both buckets are real entities. AdminFolders are ones the signed-in
    // account does not fully manage.
    const folders = [].concat(m.Folders || [], m.AdminFolders || []).map(f => ({
      id: f.Id, name: f.Name, archived: !!f.IsArchived, managed: f.CanManageFolder !== false
    }));
    const forms = [].concat(m.Forms || [], m.AdminForms || []).map(f => ({
      id: String(f.Id), name: f.Name, folderId: f.FolderId || null,
      entries: f.Entries == null ? null : f.Entries, archived: !!f.IsArchived
    }));

    const users = [];
    for (const u of roster.Users) {
      let d = null;
      try {
        d = await (await fetch('/svc/forms/permissions/' + u.Id, { credentials: 'include' })).json();
      } catch (e) { /* recorded with no overrides */ }
      users.push({
        id: u.Id, email: u.Email, first: u.FirstName || '', last: u.LastName || '',
        tier: u.Permission,
        workflowRole: (d && d.SystemRole && d.SystemRole.DisplayName) || null,
        mfa: !!u.MfaEnabled, created: u.DateCreated, verified: !!u.IsVerified,
        deleted: !!u.IsDeleted, locked: !!u.IsLocked, invitation: !!u.IsInvitation,
        folderOverrides: ((d && d.FolderRoles) || []).filter(f => f.Role && f.Role.Code)
          .map(f => ({ id: f.FolderId, code: f.Role.Code })),
        formOverrides: ((d && d.FormRoles) || []).filter(f => f.Role && f.Role.Code)
          .map(f => ({ id: String(f.FormId), code: f.Role.Code }))
      });
      await new Promise(r => setTimeout(r, 150));
    }

    // An empty formIds array omits archived forms. Naming them returns them.
    const knownForms = new Set(forms.map(f => f.id));
    const knownFolders = new Set(folders.map(f => f.id));
    const stray = new Set();
    users.forEach(u => u.formOverrides.forEach(o => knownForms.has(o.id) || stray.add(o.id)));
    if (stray.size) {
      const extra = await meta(usedOrgId, [...stray]);
      if (extra) {
        [].concat(extra.Forms || [], extra.AdminForms || []).forEach(f => {
          if (knownForms.has(String(f.Id))) return;
          knownForms.add(String(f.Id));
          forms.push({ id: String(f.Id), name: f.Name, folderId: f.FolderId || null,
                       entries: f.Entries == null ? null : f.Entries, archived: !!f.IsArchived });
        });
        [].concat(extra.Folders || [], extra.AdminFolders || []).forEach(f => {
          if (knownFolders.has(f.Id)) return;
          knownFolders.add(f.Id);
          folders.push({ id: f.Id, name: f.Name, archived: !!f.IsArchived,
                         managed: f.CanManageFolder !== false });
        });
      }
    }

    const deletedFormIds = [...stray].filter(id => !knownForms.has(id));
    const deletedFolderIds = [];
    users.forEach(u => u.folderOverrides.forEach(o => {
      if (!knownFolders.has(o.id) && deletedFolderIds.indexOf(o.id) < 0) deletedFolderIds.push(o.id);
    }));

    // Activity log. Pages with a ContinuationToken, pageSize caps at 1000.
    // Rows are aggregated as they arrive and never stored raw.
    const LOGIN_RE = /^(log(ged)?\\s?in|login|logon|sign(ed)?\\s?in|signin)$/i;
    const byEmail = {}, integrations = {}, vocab = {};
    const profileToEmail = {}, connectEvents = [];
    let logRows = 0, pages = 0, oldest = null, newest = null, truncated = false;
    try {
      let token = null;
      do {
        const b = { pageSize: 1000,
          filter: { UserId:'', ActionName:'', ItemType:'', ItemId:'', FromDate:'', ToDate:'' } };
        if (token) b.continuationToken = token;
        const r = await fetch('/svc/organization/org-audit-log/query-audit-log-page', {
          method:'POST', headers: metaHeaders(usedOrgId), credentials:'include',
          body: JSON.stringify(b) });
        if (!r.ok) break;
        const page = await r.json();
        const items = page.Items || [];
        for (const it of items) {
          logRows++;
          const when = it.Date || null;
          if (when) { if (!oldest || when < oldest) oldest = when;
                      if (!newest || when > newest) newest = when; }
          const action = it.Action || 'Unknown';
          vocab[action] = (vocab[action] || 0) + 1;
          if (it.ItemType === 'Users' && it.ItemId && it.UserEmail) {
            profileToEmail[String(it.ItemId)] = it.UserEmail.toLowerCase();
          }
          if (/connect/i.test(action) && /integration/i.test(action + ' ' + (it.ItemType || ''))) {
            connectEvents.push({ name: it.ItemName || null, by: it.UserEmail || null,
                                 when, id: it.IntegrationId || it.ItemId || null });
          }
          const key = (it.UserEmail || '').toLowerCase();
          if (key) {
            const a = byEmail[key] || (byEmail[key] = { userId: it.UserId || null,
              name: it.UserName || null, total: 0, lastActivity: null, lastLogin: null,
              logins: 0, actions: {} });
            a.total++;
            if (!a.lastActivity || when > a.lastActivity) a.lastActivity = when;
            const slot = a.actions[action] || (a.actions[action] = { n: 0, last: null });
            slot.n++;
            if (!slot.last || when > slot.last) slot.last = when;
            if (LOGIN_RE.test(action)) { a.logins++;
              if (!a.lastLogin || when > a.lastLogin) a.lastLogin = when; }
          }
          if (it.IntegrationId || it.IntegrationName) {
            const ik = it.IntegrationId || it.IntegrationName;
            const g = integrations[ik] || (integrations[ik] = { id: it.IntegrationId || null,
              name: it.IntegrationName || null, provider: it.IntegrationProvider || null,
              n: 0, last: null });
            g.n++;
            if (!g.last || when > g.last) g.last = when;
          }
        }
        token = page.ContinuationToken || null;
        pages++;
        if (pages >= 300) { truncated = true; break; }
        if (!items.length) break;
        await new Promise(r => setTimeout(r, 120));
      } while (token);
    } catch (e) { /* report runs without it */ }

    const activity = logRows ? {
      window: { from: oldest, to: newest }, rows: logRows, truncated,
      actions: Object.entries(vocab).sort((a,b) => b[1]-a[1])
        .map(([name,n]) => ({ name, n, countedAsLogin: LOGIN_RE.test(name) })),
      byEmail, integrations: Object.values(integrations), connectEvents
    } : null;

    // The log only shows integrations that have acted. A connection that has
    // never been used produces no rows, so the inventory comes from settings.
    let inventory = null;
    try {
      const inv = await (await fetch('/svc/integrations/list-view', {
        headers: metaHeaders(usedOrgId), credentials: 'include' })).json();
      if (inv && Array.isArray(inv.Integrations)) {
        const byName = {};
        connectEvents.forEach(c => {
          if (!c.name) return;
          const prev = byName[c.name];
          if (!prev || (c.when && c.when < prev.when)) byName[c.name] = c;
        });
        inventory = inv.Integrations.map(g => {
          const id = String(g.Id || '');
          const tilde = id.indexOf('~');            // appId~userProfileId for OAuth
          const appId = tilde < 0 ? id : id.slice(0, tilde);
          const owner = tilde < 0 ? null : id.slice(tilde + 1);
          const act = integrations[id] || integrations[appId] || null;
          const conn = byName[g.Name] || null;
          return { id, appId, service: g.Service || null, name: g.Name || null,
                   description: g.Description || null, userScoped: tilde >= 0,
                   ownerProfileId: owner,
                   ownerEmail: owner ? (profileToEmail[owner] || null) : null,
                   connectedBy: conn ? conn.by : null, connectedAt: conn ? conn.when : null,
                   events: act ? act.n : 0, lastSeen: act ? act.last : null };
        });
      }
    } catch (e) { /* report runs without it */ }

    return {
      schema: 4,
      activity,
      integrations: inventory,
      org: roster.OrgName || orgCode,
      orgCode,
      collectedAt: new Date().toISOString(),
      seatMessage: roster.UserUsageMessage || '',
      globalTiers: ['Administrator', 'Owner'],
      folders, forms, users, deletedFormIds, deletedFolderIds
    };
  })();
}
"""


def main():
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--org", required=True,
                    help="org code from your Cognito URL, e.g. acme in /acme/home")
    ap.add_argument("--org-id", default=None,
                    help="x-organization header value, used only if auto-detect fails")
    ap.add_argument("--timeout", type=int, default=300,
                    help="seconds to wait for sign-in (default: 300)")
    ap.add_argument("--no-open", action="store_true",
                    help="write the data file but don't open the report")
    args = ap.parse_args()

    home = f"https://www.cognitoforms.com/{args.org}/home"

    with sync_playwright() as p:
        ctx = p.chromium.launch_persistent_context(
            user_data_dir=str(PROFILE_DIR),
            headless=False,
            viewport={"width": 1280, "height": 900},
        )
        page = ctx.pages[0] if ctx.pages else ctx.new_page()
        page.goto(home, wait_until="domcontentloaded")

        if "/login" in page.url or "signin" in page.url.lower():
            print("Sign in to Cognito Forms in the window that opened.")
            print(f"Waiting up to {args.timeout}s, then collection starts on its own.")
            page.wait_for_url(f"**/{args.org}/**", timeout=args.timeout * 1000)

        print("Collecting. Roughly 15 seconds per 100 accounts.")
        page.set_default_timeout(max(args.timeout, 600) * 1000)
        payload = page.evaluate(COLLECT % json.dumps(args.org_id))
        ctx.close()

    OUT_FILE.write_text(
        "window.__AUDIT = " + json.dumps(payload, indent=2) + ";\n",
        encoding="utf-8",
    )

    users = payload["users"]
    active = [f for f in payload["forms"] if not f["archived"]]
    archived = len(payload["forms"]) - len(active)
    org_wide = [u for u in users if u["tier"] in payload["globalTiers"]]
    stale = len(payload["deletedFormIds"]) + len(payload["deletedFolderIds"])

    print(f"{len(users)} accounts, {len(active)} active forms "
          f"({archived} archived), {len(payload['folders'])} folders")
    print(f"{len(org_wide)} accounts reach every form by tier "
          f"({', '.join(sorted({u['tier'] for u in org_wide}))})")

    act = payload.get("activity")
    if act:
        logins = [a for a in act["actions"] if a["countedAsLogin"]]
        print(f"{act['rows']} log rows back to {act['window']['from'][:10]}")
        if logins:
            never = sum(1 for u in users
                        if act["byEmail"].get(u["email"].lower(), {}).get("logins", 0) == 0)
            print(f"{never} accounts have no sign-in event in that window")
        else:
            print("WARNING: no action matched as a sign-in. Check the Activity tab "
                  "in the report and adjust LOGIN_RE.")
    else:
        print("activity log unavailable; report runs without idle data")

    inv = payload.get("integrations") or []
    if inv:
        ai = [g for g in inv if (g.get("service") or "").lower() in
              ("chatgpt", "claude", "gemini", "copilot", "perplexity", "openai", "anthropic")]
        quiet = [g for g in inv if not g.get("events")]
        print(f"{len(inv)} integrations, {len(quiet)} with no recorded activity")
        if ai:
            print(f"{len(ai)} AI assistant connections: "
                  + ", ".join(g.get("ownerEmail") or g["name"] for g in ai))
    if stale:
        print(f"{stale} overrides point at deleted entities")
    print(f"wrote {OUT_FILE}")

    if not args.no_open and REPORT.exists():
        webbrowser.open(REPORT.as_uri())
        print("opened report.html")


if __name__ == "__main__":
    main()

File: collect.js
/*
 * Cognito Forms access collector
 *
 * Run from a tab signed in to Cognito Forms, on the org home page
 * (https://www.cognitoforms.com/<yourorg>/home).
 *
 * Cognito's permission model is a global tier plus sparse overrides. This
 * collects both: the roster carries each account's tier, and the per-user
 * endpoint carries only the folders and forms where someone changed it.
 * Codes are AD (Administrator), ED (Editor), RE (Reviewer), NO (denied).
 * A folder with no row inherits the tier.
 *
 * Everything here is a read. Nothing is written back to Cognito Forms.
 */
(async () => {
  // Fallback only, and normally unnecessary. Set this to your x-organization
  // header value if the form list comes back empty. See the README.
  const CONFIG = { orgId: null };

  /* ---------------- overlay ---------------- */
  document.getElementById('cognito-audit-ui')?.remove();
  const host = document.createElement('div');
  host.id = 'cognito-audit-ui';
  host.style.cssText = 'position:fixed;right:20px;bottom:20px;z-index:2147483647';
  document.body.appendChild(host);
  const root = host.attachShadow({ mode: 'open' });
  root.innerHTML = [
    '<style>',
    '.card{width:330px;background:#131c24;color:#e7ecf0;border:1px solid #2b3a47;',
    'font:13px/1.45 ui-monospace,SFMono-Regular,Menlo,monospace;',
    'box-shadow:0 12px 32px rgba(0,0,0,.35)}',
    '.hd{display:flex;justify-content:space-between;align-items:center;',
    'padding:10px 12px;border-bottom:1px solid #2b3a47}',
    '.hd b{font-weight:500;font-size:12px;letter-spacing:.02em}',
    '.x{background:none;border:0;color:#7d919f;cursor:pointer;font-size:16px;line-height:1;padding:0 2px}',
    '.x:hover{color:#e7ecf0}',
    '.bd{padding:12px}',
    '.step{color:#9fb3c2;margin-bottom:10px;min-height:2.9em;word-break:break-all}',
    '.track{height:4px;background:#22303c;overflow:hidden}',
    '.fill{height:100%;width:0;background:#4a9ed4;transition:width .18s}',
    '.fill.done{background:#4fae83}',
    '.fill.err{background:#d1604a}',
    '.nums{display:flex;justify-content:space-between;color:#7d919f;font-size:11px;padding-top:8px}',
    '.act{margin-top:12px;display:none}',
    '.act button{width:100%;background:#1f6da1;color:#fff;border:0;padding:8px;cursor:pointer;font:inherit}',
    '.act button:hover{background:#17567f}',
    '.note{color:#7d919f;font-size:11px;margin-top:8px}',
    '.err{color:#f0a294}',
    '</style>',
    '<div class="card">',
    '<div class="hd"><b>COGNITO ACCESS COLLECTOR</b><button class="x" title="Close">&times;</button></div>',
    '<div class="bd">',
    '<div class="step" id="step">Starting.</div>',
    '<div class="track"><div class="fill" id="fill"></div></div>',
    '<div class="nums"><span id="left"></span><span id="right"></span></div>',
    '<div class="act" id="act"><button id="again">Save another copy</button></div>',
    '<div class="note" id="note"></div>',
    '</div></div>'
  ].join('');

  const el = s => root.getElementById(s);
  root.querySelector('.x').onclick = () => host.remove();

  const ui = {
    step: t => { el('step').innerHTML = t; },
    bar: (p, c) => { el('fill').style.width = p + '%'; el('fill').className = 'fill ' + (c || ''); },
    nums: (l, r) => { el('left').textContent = l || ''; el('right').textContent = r || ''; },
    note: t => { el('note').innerHTML = t || ''; },
    fail: t => { el('step').innerHTML = '<span class="err">' + t + '</span>'; ui.bar(100, 'err'); }
  };

  /* ---------------- guard ---------------- */
  const orgCode = location.pathname.split('/')[1] || '';
  if (!/cognitoforms\.com$/.test(location.hostname) || !orgCode) {
    ui.fail('Run this from your Cognito Forms org page, e.g. /yourorg/home');
    return;
  }

  /* ---------------- save location ----------------
     showSaveFilePicker needs a recent click, so ask now and write at the end. */
  let handle = null;
  if (typeof window.showSaveFilePicker === 'function') {
    ui.step('Choose where to save <b>audit-data.js</b>.<br>Collection starts as soon as you pick.');
    ui.bar(4);
    try {
      handle = await window.showSaveFilePicker({
        suggestedName: 'audit-data.js',
        types: [{ description: 'Audit data', accept: { 'text/javascript': ['.js'] } }]
      });
    } catch (e) {
      if (e && e.name === 'AbortError') {
        ui.fail('Cancelled.');
        setTimeout(() => host.remove(), 2000);
        return;
      }
      handle = null;
    }
  }

  async function save(text) {
    if (handle) {
      const w = await handle.createWritable();
      await w.write(text);
      await w.close();
      return handle.name;
    }
    const a = document.createElement('a');
    a.href = URL.createObjectURL(new Blob([text], { type: 'text/javascript' }));
    a.download = 'audit-data.js';
    a.style.display = 'none';
    document.body.appendChild(a);
    a.click();
    setTimeout(() => { URL.revokeObjectURL(a.href); a.remove(); }, 4000);
    return 'audit-data.js, in your downloads folder';
  }

  const META_HEADERS = orgId => {
    const h = { 'content-type': 'application/json', 'x-requested-with': 'XMLHttpRequest' };
    if (orgId) h['x-organization'] = orgId;
    return h;
  };
  async function meta(orgId, formIds) {
    const r = await fetch('/svc/entity-meta/forms', {
      method: 'POST', headers: META_HEADERS(orgId), credentials: 'include',
      body: JSON.stringify({ formIds: formIds || [] })
    });
    if (!r.ok) return null;
    try { return await r.json(); } catch (e) { return null; }
  }

  /* ---------------- 1. roster ---------------- */
  ui.step('Reading the account roster.');
  ui.bar(6);
  let roster;
  try {
    roster = await (await fetch('/svc/organization/users', { credentials: 'include' })).json();
  } catch (e) {
    ui.fail('Could not read the roster. Are you still signed in?');
    return;
  }
  if (!Array.isArray(roster.Users) || !roster.Users.length) {
    ui.fail('The roster came back empty. Sign in again and retry.');
    return;
  }
  ui.nums(roster.Users.length + ' accounts', '');

  /* ---------------- 2. active forms and folders ---------------- */
  ui.step('Reading forms and folders.');
  ui.bar(10);
  let m = await meta(null, []);
  let usedOrgId = null;
  if (!m || !(m.Forms || []).length) { m = await meta(CONFIG.orgId, []); usedOrgId = CONFIG.orgId; }
  if (!m || !(m.Forms || []).length) {
    ui.fail('The form list came back empty. Check CONFIG.orgId against the x-organization header on the entity-meta/forms request.');
    return;
  }

  // Folders and forms arrive in two buckets. The admin ones are entities the
  // signed-in account does not fully manage. Both are real; merge them.
  const folders = [].concat(m.Folders || [], m.AdminFolders || []).map(f => ({
    id: f.Id, name: f.Name, archived: !!f.IsArchived, managed: f.CanManageFolder !== false
  }));
  const forms = [].concat(m.Forms || [], m.AdminForms || []).map(f => ({
    id: String(f.Id), name: f.Name, folderId: f.FolderId || null,
    entries: f.Entries == null ? null : f.Entries, archived: !!f.IsArchived
  }));
  ui.nums(roster.Users.length + ' accounts', forms.length + ' forms · ' + folders.length + ' folders');

  /* ---------------- 3. per-user overrides ----------------
     Every row is kept, including NO. On a tier that already grants access, a
     NO row is a deliberate denial, and dropping it overstates reach. */
  const total = roster.Users.length;
  const users = [];
  let n = 0, failed = 0;

  for (const u of roster.Users) {
    ui.step('Reading overrides.<br>' + u.Email);
    let d = null;
    try {
      d = await (await fetch('/svc/forms/permissions/' + u.Id, { credentials: 'include' })).json();
    } catch (e) {
      failed++;
    }
    users.push({
      id: u.Id, email: u.Email, first: u.FirstName || '', last: u.LastName || '',
      tier: u.Permission,
      workflowRole: (d && d.SystemRole && d.SystemRole.DisplayName) || null,
      mfa: !!u.MfaEnabled, created: u.DateCreated, verified: !!u.IsVerified,
      deleted: !!u.IsDeleted, locked: !!u.IsLocked, invitation: !!u.IsInvitation,
      folderOverrides: ((d && d.FolderRoles) || [])
        .filter(f => f.Role && f.Role.Code)
        .map(f => ({ id: f.FolderId, code: f.Role.Code })),
      formOverrides: ((d && d.FormRoles) || [])
        .filter(f => f.Role && f.Role.Code)
        .map(f => ({ id: String(f.FormId), code: f.Role.Code }))
    });
    n++;
    ui.bar(10 + Math.round(n / total * 78));
    ui.nums(n + ' of ' + total, failed ? failed + ' failed' : forms.length + ' forms');
    await new Promise(r => setTimeout(r, 150));
  }

  /* ---------------- 4. resolve archived entities ----------------
     An empty formIds array omits archived forms, but naming them explicitly
     returns them. Overrides routinely point at archived forms, so without
     this pass they show up as unresolvable references. */
  ui.step('Resolving archived forms.');
  ui.bar(92);
  const knownForms = new Set(forms.map(f => f.id));
  const knownFolders = new Set(folders.map(f => f.id));
  const strayForms = new Set();
  users.forEach(u => u.formOverrides.forEach(o => knownForms.has(o.id) || strayForms.add(o.id)));

  let archivedAdded = 0;
  if (strayForms.size) {
    const extra = await meta(usedOrgId, [...strayForms]);
    if (extra) {
      [].concat(extra.Forms || [], extra.AdminForms || []).forEach(f => {
        if (knownForms.has(String(f.Id))) return;
        knownForms.add(String(f.Id));
        archivedAdded++;
        forms.push({
          id: String(f.Id), name: f.Name, folderId: f.FolderId || null,
          entries: f.Entries == null ? null : f.Entries, archived: !!f.IsArchived
        });
      });
      [].concat(extra.Folders || [], extra.AdminFolders || []).forEach(f => {
        if (knownFolders.has(f.Id)) return;
        knownFolders.add(f.Id);
        folders.push({ id: f.Id, name: f.Name, archived: !!f.IsArchived,
                       managed: f.CanManageFolder !== false });
      });
    }
  }

  const deletedForms = [...strayForms].filter(id => !knownForms.has(id));
  const deletedFolders = [];
  users.forEach(u => u.folderOverrides.forEach(o => {
    if (!knownFolders.has(o.id) && deletedFolders.indexOf(o.id) < 0) deletedFolders.push(o.id);
  }));

  /* ---------------- 5. activity log ----------------
     org-audit-log pages with a ContinuationToken and caps pageSize at 1000.
     Rows are aggregated per account as they arrive, never stored raw, so the
     payload stays small no matter how long the retention window is. */
  ui.step('Reading the activity log.');
  ui.bar(94);

  const LOGIN_RE = /^(log(ged)?\s?in|login|logon|sign(ed)?\s?in|signin)$/i;
  const byEmail = {};
  const integrations = {};
  const actionVocab = {};
  const profileToEmail = {};
  const connectEvents = [];
  let rows = 0, pages = 0, oldest = null, newest = null, truncated = false;
  const PAGE_CAP = 300; // 300k rows, well past any plausible org

  try {
    let token = null;
    do {
      const body = { pageSize: 1000,
        filter: { UserId: '', ActionName: '', ItemType: '', ItemId: '', FromDate: '', ToDate: '' } };
      if (token) body.continuationToken = token;
      const r = await fetch('/svc/organization/org-audit-log/query-audit-log-page', {
        method: 'POST', headers: META_HEADERS(usedOrgId), credentials: 'include',
        body: JSON.stringify(body)
      });
      if (!r.ok) break;
      const page = await r.json();
      const items = page.Items || [];

      for (const it of items) {
        rows++;
        const when = it.Date || null;
        if (when) {
          if (!oldest || when < oldest) oldest = when;
          if (!newest || when > newest) newest = when;
        }
        const action = it.Action || 'Unknown';
        actionVocab[action] = (actionVocab[action] || 0) + 1;

        // Rows of ItemType 'Users' carry the subject's profile id in ItemId.
        // That id is the suffix of an OAuth integration's compound id, which is
        // how integrations get attributed to an owner.
        if (it.ItemType === 'Users' && it.ItemId && it.UserEmail) {
          profileToEmail[String(it.ItemId)] = it.UserEmail.toLowerCase();
        }
        if (/connect/i.test(action) && /integration/i.test(action + ' ' + (it.ItemType || ''))) {
          connectEvents.push({ name: it.ItemName || null, by: it.UserEmail || null,
                               when: when, id: it.IntegrationId || it.ItemId || null });
        }

        const key = (it.UserEmail || '').toLowerCase();
        if (key) {
          const a = byEmail[key] || (byEmail[key] = {
            userId: it.UserId || null, name: it.UserName || null,
            total: 0, lastActivity: null, lastLogin: null, logins: 0, actions: {} });
          a.total++;
          if (!a.lastActivity || when > a.lastActivity) a.lastActivity = when;
          const slot = a.actions[action] || (a.actions[action] = { n: 0, last: null });
          slot.n++;
          if (!slot.last || when > slot.last) slot.last = when;
          if (LOGIN_RE.test(action)) {
            a.logins++;
            if (!a.lastLogin || when > a.lastLogin) a.lastLogin = when;
          }
        }

        if (it.IntegrationId || it.IntegrationName) {
          const ik = it.IntegrationId || it.IntegrationName;
          const g = integrations[ik] || (integrations[ik] = {
            id: it.IntegrationId || null, name: it.IntegrationName || null,
            provider: it.IntegrationProvider || null, n: 0, last: null });
          g.n++;
          if (!g.last || when > g.last) g.last = when;
        }
      }

      token = page.ContinuationToken || null;
      pages++;
      ui.nums(rows + ' log rows', pages + ' pages');
      if (pages >= PAGE_CAP) { truncated = true; break; }
      if (!items.length) break;
      await new Promise(r => setTimeout(r, 120));
    } while (token);
  } catch (e) {
    console.warn('[cognito-audit] activity log unavailable', e);
  }

  /* ---------------- 5b. integration inventory ----------------
     The log only shows integrations that have acted. A connection that has
     never been used produces no rows, so the inventory has to come from the
     settings endpoint or it stays invisible. */
  let inventory = null;
  try {
    const inv = await (await fetch('/svc/integrations/list-view', {
      headers: META_HEADERS(usedOrgId), credentials: 'include'
    })).json();
    if (inv && Array.isArray(inv.Integrations)) {
      const connectByName = {};
      connectEvents.forEach(c => {
        if (!c.name) return;
        const prev = connectByName[c.name];
        if (!prev || (c.when && c.when < prev.when)) connectByName[c.name] = c;
      });

      inventory = inv.Integrations.map(g => {
        const id = String(g.Id || '');
        // OAuth app integrations use appId~userProfileId. API keys have no ~.
        const tilde = id.indexOf('~');
        const appId = tilde < 0 ? id : id.slice(0, tilde);
        const owner = tilde < 0 ? null : id.slice(tilde + 1);
        const act = integrations[id] || integrations[appId] || null;
        const conn = connectByName[g.Name] || null;
        return {
          id: id, appId: appId, service: g.Service || null, name: g.Name || null,
          description: g.Description || null,
          userScoped: tilde >= 0,
          ownerProfileId: owner,
          ownerEmail: owner ? (profileToEmail[owner] || null) : null,
          connectedBy: conn ? conn.by : null,
          connectedAt: conn ? conn.when : null,
          events: act ? act.n : 0,
          lastSeen: act ? act.last : null
        };
      });
    }
  } catch (e) {
    console.warn('[cognito-audit] integration inventory unavailable', e);
  }

  const activity = rows ? {
    window: { from: oldest, to: newest },
    rows: rows, truncated: truncated,
    // Vocabulary is reported so a mismatch in the login action name is visible
    // rather than silently producing zero logins for everyone.
    actions: Object.entries(actionVocab).sort((a, b) => b[1] - a[1])
      .map(([name, n]) => ({ name: name, n: n, countedAsLogin: LOGIN_RE.test(name) })),
    byEmail: byEmail,
    integrations: Object.values(integrations),
    connectEvents: connectEvents
  } : null;

  /* ---------------- 6. save ---------------- */
  const payload = {
    schema: 4,
    activity: activity,
    integrations: inventory,
    org: roster.OrgName || orgCode,
    orgCode: orgCode,
    collectedAt: new Date().toISOString(),
    seatMessage: roster.UserUsageMessage || '',
    // Tiers that reach every form unless an override says otherwise.
    globalTiers: ['Administrator', 'Owner'],
    folders: folders, forms: forms, users: users,
    deletedFormIds: deletedForms, deletedFolderIds: deletedFolders
  };
  const text = 'window.__AUDIT = ' + JSON.stringify(payload, null, 2) + ';\n';
  window.__AUDIT = payload;

  let where;
  try {
    where = await save(text);
  } catch (e) {
    ui.fail('Collection finished but the file could not be written. Use the button below.');
    el('act').style.display = 'block';
    el('again').onclick = () => { handle = null; save(text); };
    return;
  }

  const globals = users.filter(u => payload.globalTiers.indexOf(u.tier) >= 0).length;
  const loginActions = activity ? activity.actions.filter(a => a.countedAsLogin) : [];
  ui.bar(100, 'done');
  ui.step('Done. Saved to <b>' + where + '</b>');
  ui.nums(total + ' accounts · ' + globals + ' org-wide', failed ? failed + ' failed' : '');
  ui.note('Put this file next to report.html and open it, or drop it on the report.'
    + (activity
        ? '<br>' + activity.rows + ' log rows back to ' + String(activity.window.from).slice(0, 10)
          + (loginActions.length
              ? '' : '<br><span class="err">No sign-in action matched. Check the action list in the report.</span>')
        : '<br><span class="err">Activity log unavailable. Report runs without it.</span>')
    + (inventory ? '<br>' + inventory.length + ' integrations, '
        + inventory.filter(g => !g.events).length + ' with no recorded activity.' : '')
    + (archivedAdded ? '<br>' + archivedAdded + ' archived forms resolved.' : '')
    + (deletedForms.length + deletedFolders.length
        ? '<br>' + (deletedForms.length + deletedFolders.length) + ' overrides point at deleted entities.' : '')
    + (failed ? '<br><span class="err">Some accounts returned no data. They appear with no overrides.</span>' : ''));

  el('act').style.display = 'block';
  el('again').onclick = () => { handle = null; save(text); };
})();
